tensorizer/stream_io.py

```diff
@@ -417,6 +422,292 @@ def seek(self, position, whence=SEEK_SET):
+
+    if uri_components.scheme.lower() != "redis":
+        raise ValueError(f"Invalid Redis URI: {uri}")
+
+    host = uri_components.hostname
+    port = uri_components.port
+    if port is None:
+        port = 6379
+    prefix = uri_components.path.lstrip("/")
+
+    return host, port, prefix
+
+
+# Detect if we're running on OSX, and if so, set max buffer size to 1 MiB.
+if sys.platform == "darwin":
+    _MAX_TCP_BUFFER_SIZE = 2 << 20  # 1 MiB if OSX
+else:
+    _MAX_TCP_BUFFER_SIZE = 2 << 22  # 8 MiB
+
+
+class RedisStreamFile:
+    """
+    RedisStreamFile implements a file-like object around a Redis key namespace. Each
+    'file' is broken up into multiple keys, each of which is a slice of the file. Each
+    key is named with the prefix, followed by a colon, a user-assigned name, another
+    colon, and then the byte index of the key.
+
+    For example, if the prefix is 'foo', and the user-assigned name is 'bar', then the
+    keys would be 'foo:bar:0', 'foo:bar:16'. The first key would be the first 16 bytes
+    of the file, and the remainder would be in the next key.
+
+    On initialization, it performs a prefix scan of the keys for the prefix, and then
+    sorts the keys by their byte index. This allows it to perform a seek operation
+    efficiently, as it can find the key that contains the byte index, and then read
+    from that key.
+
+    RedisStreamFile has some optimizations, such as reading the entirety of a key into
+    a buffer, and then serving subsequent reads from that buffer. It also has a buffer
+    for the remainder of a key, so that it can serve partial reads of a key.
+
+    Arguments:
+        uri: The URI of the Redis key namespace. This should be prefixed by the
+            'redis://' scheme, and include the key prefix. For example,
+            'redis://localhost:6379/foo'.
+        buffer_size: The size of the TCP buffer to use for the Redis connection.
+
+    Attributes:
+        setup_latency: The time it took to setup the Redis connections and enumerate.
+        response_latencies: A list of the time it took to get the Redis responses.
+        bytes_read: The number of bytes read from the stream.
+        bytes_skipped: The number of bytes skipped from the stream.
+        read_operations: The number of read operations performed on the stream.
+    """
+
+    def __init__(
+        self,
+        uri: str,
+        *,
+        buffer_size: int = _MAX_TCP_BUFFER_SIZE,
+    ) -> None:
+        if buffer_size is None:
+            buffer_size = _MAX_TCP_BUFFER_SIZE
+        init_begin = time.monotonic()
+        host, port, prefix = _parse_redis_uri(uri)
+        self._redis = redis.Redis(host=host, port=port, db=0)
+        self._redis_tcp = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self._redis_tcp.setsockopt(
+            socket.SOL_SOCKET, socket.SO_RCVBUF, buffer_size
+        )
+        self._redis_tcp.connect((host, port))
+        self._redis_tcp_mutex = threading.Lock()
```

Comment:
Multithreading is not used in this file's own code; what is the justification for guarding `_read_from` with a mutex? Concurrent `read` calls on file objects are usually already considered unsafe and protected by locks in the caller's code if they desire concurrent access, especially in our own `serialization.py`.

Reply:
Internal for overlapping buffering and I/O. External reads do not mean internal reads.